### PR TITLE
Remove synchronous calls that can block from TV code

### PR DESCRIPTION
**Pull request by chrisdecenzo**
A number of commands and message processing paths can involve blocking tasks, including:
1. content app interactions (command and attribute reads)
2. loading apps

These need to be migrated to async calls with callbacks

### CODE REVIEW

File: src/controller/CommissionerDiscoveryController.cpp [hunk 1]
```diff
@@ -112,6 +112,8 @@ void CommissionerDiscoveryController::OnUserDirectedCommissioningRequest(UDCClie
 
 void CommissionerDiscoveryController::Ok()
 {
+    chip::DeviceLayer::StackLock lock;
```
Comment (bzbarsky-apple) [inline src/controller/CommissionerDiscoveryController.cpp:115]: This is assuming it's running on a setup where StackLock is not a no-op, no?  That's not true for controller platforms in general.

The right way to run things that need to be run in the Matter context is via async dispatch to the Matter event loop, if you want it to work across all platforms.

File: src/controller/CommissionerDiscoveryController.cpp [hunk 2]
```diff
@@ -135,55 +137,163 @@ void CommissionerDiscoveryController::Ok()
...
-        // Encoding::BytesToUppercaseHexString(client->GetRotatingId(), chip::Dnssd::kMaxRotatingIdLen, rotatingIdString,
-        //                                     sizeof(rotatingIdString));
+        HandleContentAppPasscodeResponse(0);
+        return;
+    }
+
+    char rotatingIdString[chip::Dnssd::kMaxRotatingIdLen * 2 + 1] = "";
+    Encoding::BytesToUppercaseHexString(client->GetRotatingId(), client->GetRotatingIdLength(), rotatingIdString,
+                                        sizeof(rotatingIdString));
+    CharSpan rotatingIdSpan = chip::CharSpan(rotatingIdString, sizeof(rotatingIdString));
```
Comment (bzbarsky-apple) [inline src/controller/CommissionerDiscoveryController.cpp:149]: This size looks wrong.

(And separately, why the extra chip:: prefix?)

File: src/controller/CommissionerDiscoveryController.cpp [hunk 3]
```diff
@@ -135,55 +137,163 @@ void CommissionerDiscoveryController::Ok()
...
+        return;
+    }
+
+    mPasscodeService->FetchCommissionPasscodeFromContentApp(client->GetVendorId(), client->GetProductId(), rotatingIdSpan);
+    return;
+}
+
+void CommissionerDiscoveryController::HandleTargetContentAppCheck(TargetAppInfo target, uint32_t passcode)
+{
+    chip::DeviceLayer::StackLock lock;
```
Comment (bzbarsky-apple) [inline src/controller/CommissionerDiscoveryController.cpp:171]: Again, you should probably not be using StackLock.

File: src/controller/CommissionerDiscoveryController.cpp (hunk 1 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: I updated these commands to spawn a task on the main chip thread.

File: src/controller/CommissionerDiscoveryController.cpp (hunk 2 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: removed the prefix. What about the size does not look right?

File: src/controller/CommissionerDiscoveryController.cpp (hunk 3 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: removed the StackLock calls

**Review by lazarkov — APPROVED** [review]

File: src/controller/CommissionerDiscoveryController.cpp (hunk 2 above)
Comment (bzbarsky-apple) [reply src/controller/CommissionerDiscoveryController.cpp]: Well, what will be contained in the span given the size that's being used?

File: src/controller/CommissionerDiscoveryController.cpp [hunk 4]
```diff
@@ -34,6 +34,13 @@
 using namespace ::chip;
 using namespace chip::Protocols::UserDirectedCommissioning;
 
+CommissionerDiscoveryController gCommissionerDiscoveryController;
```
Comment (bzbarsky-apple) [inline src/controller/CommissionerDiscoveryController.cpp:37]: So this forces on the core SDK level (unlike before, where it was in the example app), that there is a unique commissioner discovery controller.  Is that reasonable?  I don't have a good feel for how this gets used in practice by SDK consumers.

File: src/controller/CommissionerDiscoveryController.cpp [hunk 5]
```diff
@@ -43,8 +50,28 @@ void CommissionerDiscoveryController::ResetState()
...
+    if (mUdcServer != nullptr)
+    {
+        UDCClientState * client = mUdcServer->GetUDCClients().FindUDCClientState(mCurrentInstance);
+        if (client != nullptr)
+        {
+            // everything looks good
+            return;
+        }
+    }
+    ResetState();
```
Comment (bzbarsky-apple) [inline src/controller/CommissionerDiscoveryController.cpp:68]: This sounds like a side effect, but CheckValidSession() kind of sounds like a side-effect-free thing that should return a boolean.

What is this function actually doing?  It feels mis-named.

File: src/controller/CommissionerDiscoveryController.cpp [hunk 6]
```diff
@@ -94,24 +120,73 @@ void CommissionerDiscoveryController::OnUserDirectedCommissioningRequest(UDCClie
     mReady = false;
     Platform::CopyString(mCurrentInstance, state.GetInstanceName());
     mPendingConsent = true;
-    char rotatingDeviceIdHexBuffer[RotatingDeviceId::kHexMaxLength];
-    Encoding::BytesToUppercaseHexString(state.GetRotatingId(), state.GetRotatingIdLength(), rotatingDeviceIdHexBuffer,
-                                        RotatingDeviceId::kHexMaxLength);
+    char rotatingIdString[Dnssd::kMaxRotatingIdLen * 2 + 1];
```
Comment (bzbarsky-apple) [inline src/controller/CommissionerDiscoveryController.cpp:123]: So is there a reason this is no longer using `RotatingDeviceId::kHexMaxLength`?  If there is, there should probably some documentation where that constant is defined explaining what it means, because on its face it seems like it would be usable here.

File: src/controller/CommissionerDiscoveryController.cpp [hunk 7]
```diff
@@ -94,24 +120,73 @@ void CommissionerDiscoveryController::OnUserDirectedCommissioningRequest(UDCClie
...
         mUserPrompter->PromptForCommissionOKPermission(state.GetVendorId(), state.GetProductId(), state.GetDeviceName());
     }
     ChipLogDetail(Controller, "------Via Shell Enter: controller ux ok|cancel");
 }
 
+/// Callback for getting execution into the main chip thread
+void CallbackOk(System::Layer * aSystemLayer, void * aAppState)
+{
+    ChipLogError(AppServer, "UX Ok: now on main thread");
+    GetCommissionerDiscoveryController()->InternalOk();
```
Comment (bzbarsky-apple) [inline src/controller/CommissionerDiscoveryController.cpp:143]: Shouldn't we just pass the controller via aAppState instead of relying on globals?

File: src/controller/CommissionerDiscoveryController.cpp [hunk 8]
```diff
@@ -94,24 +120,73 @@ void CommissionerDiscoveryController::OnUserDirectedCommissioningRequest(UDCClie
...
+
+    GetCommissionerDiscoveryController()->GetPasscodeService()->FetchCommissionPasscodeFromContentApp(
+        client->GetVendorId(), client->GetProductId(), rotatingIdSpan);
+}
+
 void CommissionerDiscoveryController::Ok()
 {
+    ChipLogError(AppServer, "UX Ok: moving to main thread");
+    // need to ensure callback is on main chip thread
+    DeviceLayer::SystemLayer().StartTimer(System::Clock::Seconds32(0), CallbackOk, nullptr);
```
Comment (bzbarsky-apple) [inline src/controller/CommissionerDiscoveryController.cpp:181]: This doesn't make sense.  StartTimer is clearly documented as:
```
This method must be called while in the Matter context (from
the Matter event loop, or while holding the Matter stack lock).
```

and asserts that fact.  I have to ask: was this tested?

File: src/controller/CommissionerDiscoveryController.cpp [hunk 9]
```diff
@@ -94,24 +120,73 @@ void CommissionerDiscoveryController::OnUserDirectedCommissioningRequest(UDCClie
...
+    ChipLogError(AppServer, "UX Ok: now on main thread");
+    GetCommissionerDiscoveryController()->InternalOk();
+}
+
+/// Callback for getting execution out of the main chip thread
+void CallbackExternalWorkerOk(System::Layer * aSystemLayer, void * aAppState)
+{
+    ChipLogError(AppServer, "UX Ok: out of main thread");
+
+    UDCClientState * client = static_cast<UDCClientState *>(aAppState);
```
Comment (bzbarsky-apple) [inline src/controller/CommissionerDiscoveryController.cpp:151]: Does something guarantee that client state does not get overwritten while the callback was pending, or while it's now executed?  Last I checked, FindUDCClientState would happily return you a pointer to a state just based on time elapsed, so if things take too long somewhere someone can have stomped on the client state, no?

File: src/controller/CommissionerDiscoveryController.cpp [hunk 10]
```diff
@@ -94,24 +120,73 @@ void CommissionerDiscoveryController::OnUserDirectedCommissioningRequest(UDCClie
...
+
+    char rotatingIdString[Dnssd::kMaxRotatingIdLen * 2 + 1] = "";
+    Encoding::BytesToUppercaseHexString(client->GetRotatingId(), client->GetRotatingIdLength(), rotatingIdString,
+                                        sizeof(rotatingIdString));
+    CharSpan rotatingIdSpan = CharSpan(rotatingIdString, sizeof(rotatingIdString));
```
Comment (bzbarsky-apple) [inline src/controller/CommissionerDiscoveryController.cpp:156]: Again, this size looks wrong.

File: src/controller/CommissionerDiscoveryController.cpp [hunk 11]
```diff
@@ -94,24 +120,73 @@ void CommissionerDiscoveryController::OnUserDirectedCommissioningRequest(UDCClie
...
 void CommissionerDiscoveryController::Ok()
 {
+    ChipLogError(AppServer, "UX Ok: moving to main thread");
+    // need to ensure callback is on main chip thread
+    DeviceLayer::SystemLayer().StartTimer(System::Clock::Seconds32(0), CallbackOk, nullptr);
+}
+
+void CommissionerDiscoveryController::InternalOk()
+{
+    ChipLogError(AppServer, "UX InternalOk");
```
Comment (bzbarsky-apple) [inline src/controller/CommissionerDiscoveryController.cpp:186]: Should `assertChipStackLockedByCurrentThread()`, yes?  And include the right header for that.

File: src/controller/CommissionerDiscoveryController.cpp [hunk 12]
```diff
@@ -135,55 +210,178 @@ void CommissionerDiscoveryController::Ok()
...
+    GetCommissionerDiscoveryController()->InternalHandleTargetContentAppCheck(*target, target->passcode);
+    delete target;
+}
+
+void CommissionerDiscoveryController::HandleTargetContentAppCheck(TargetAppInfo target, uint32_t passcode)
+{
+    // need to ensure callback is on main chip thread
+    target.passcode      = passcode;
+    TargetAppInfo * info = new TargetAppInfo(target);
+    DeviceLayer::SystemLayer().StartTimer(System::Clock::Seconds32(0), CallbackHandleTargetContentAppCheck, info);
```
Comment (bzbarsky-apple) [inline src/controller/CommissionerDiscoveryController.cpp:240]: StartTimer can fail, and then you have a memory leak.

But more importantly: you can't call StartTimer if you're not on the Matter thread or holding the Matter lock.

File: src/controller/CommissionerDiscoveryController.cpp [hunk 13]
```diff
@@ -135,55 +210,178 @@ void CommissionerDiscoveryController::Ok()
...
+}
+
+void CommissionerDiscoveryController::InternalHandleTargetContentAppCheck(TargetAppInfo target, uint32_t passcode)
+{
+    CheckValidSession();
```
Comment (bzbarsky-apple) [inline src/controller/CommissionerDiscoveryController.cpp:245]: Should `assertChipStackLockedByCurrentThread()`

File: src/controller/CommissionerDiscoveryController.cpp [hunk 14]
```diff
@@ -135,55 +210,178 @@ void CommissionerDiscoveryController::Ok()
...
+                }
             }
         }
-        else
-        {
-            passcode = mPasscodeService->FetchCommissionPasscodeFromContentApp(client->GetVendorId(), client->GetProductId(),
-                                                                               rotatingIdSpan);
-        }
+    }
+    if (passcode != 0)
```
Comment (bzbarsky-apple) [inline src/controller/CommissionerDiscoveryController.cpp:310]: I did not review the logic changes here carefully.  I'm not sure why there are logic changes here at all; they seem unrelated to teh "remove sync calls" bit.

File: src/controller/CommissionerDiscoveryController.cpp [hunk 15]
```diff
@@ -135,55 +210,178 @@ void CommissionerDiscoveryController::Ok()
...
+{
+    uint32_t * tmpPasscode = static_cast<uint32_t *>(aAppState);
+    GetCommissionerDiscoveryController()->InternalHandleContentAppPasscodeResponse(*tmpPasscode);
+    delete tmpPasscode;
+}
+
+void CommissionerDiscoveryController::HandleContentAppPasscodeResponse(uint32_t passcode)
+{
+    uint32_t * tmpPasscode = new uint32_t(passcode);
+    DeviceLayer::SystemLayer().StartTimer(System::Clock::Seconds32(0), CallbackHandleContentAppPasscodeResponse,
```
Comment (bzbarsky-apple) [inline src/controller/CommissionerDiscoveryController.cpp:348]: All the same comments as above.

File: src/controller/CommissionerDiscoveryController.cpp [hunk 16]
```diff
@@ -259,8 +455,25 @@ void CommissionerDiscoveryController::Ok()
...
+{
+    uint32_t * tmpPasscode = static_cast<uint32_t *>(aAppState);
+    GetCommissionerDiscoveryController()->InternalCommissionWithPasscode(*tmpPasscode);
+    delete tmpPasscode;
+}
+
 void CommissionerDiscoveryController::CommissionWithPasscode(uint32_t passcode)
 {
+    uint32_t * tmpPasscode = new uint32_t(passcode);
+    DeviceLayer::SystemLayer().StartTimer(System::Clock::Seconds32(0), CallbackCommissionWithPasscode,
```
Comment (bzbarsky-apple) [inline src/controller/CommissionerDiscoveryController.cpp:469]: Same comments as above.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/controller/CommissionerDiscoveryController.h [hunk 17]
```diff
@@ -273,18 +276,42 @@ class CommissionerDiscoveryController : public chip::Protocols::UserDirectedComm
...
+     */
+    void HandleContentAppPasscodeResponse(uint32_t passcode);
+    void InternalHandleContentAppPasscodeResponse(uint32_t passcode);
+
+    /**
+     * @brief
+     *   Called with the result of attempting to lookup and obtain the passcode from the content app corresponding to the given
+     * target.
+     *
+     *  @param[in]    passcode           Passcode for the given commissionee, or 0 if passcode cannot be obtained.
```
Comment (yunhanw-google) [inline src/controller/CommissionerDiscoveryController.h:303]: missing target input parameter.

File: src/controller/CommissionerDiscoveryController.cpp [hunk 18]
```diff
@@ -259,8 +473,32 @@ void CommissionerDiscoveryController::Ok()
...
+void CallbackCommissionWithPasscode(System::Layer * aSystemLayer, void * aAppState)
+{
+    uint32_t * tmpPasscode = static_cast<uint32_t *>(aAppState);
+    GetCommissionerDiscoveryController()->InternalCommissionWithPasscode(*tmpPasscode);
+    delete tmpPasscode;
+}
+
 void CommissionerDiscoveryController::CommissionWithPasscode(uint32_t passcode)
 {
+    uint32_t * tmpPasscode = new uint32_t(passcode);
```
Comment (yunhanw-google) [inline src/controller/CommissionerDiscoveryController.cpp:486]: check whether tmpPasscode is allocated successfully.

File: src/controller/CommissionerDiscoveryController.cpp [hunk 19]
```diff
@@ -259,8 +473,32 @@ void CommissionerDiscoveryController::Ok()
     ChipLogDetail(Controller, "------Via Shell Enter: controller ux ok [passcode]");
 }
 
+/// Callback for getting execution into the main chip thread
+void CallbackCommissionWithPasscode(System::Layer * aSystemLayer, void * aAppState)
+{
+    uint32_t * tmpPasscode = static_cast<uint32_t *>(aAppState);
```
Comment (yunhanw-google) [inline src/controller/CommissionerDiscoveryController.cpp:479]: check whether it is nullptr

File: src/controller/CommissionerDiscoveryController.cpp (hunk 4 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: reverted

File: src/controller/CommissionerDiscoveryController.cpp (hunk 7 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: fixed

File: src/controller/CommissionerDiscoveryController.cpp (hunk 8 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: fixed

File: src/controller/CommissionerDiscoveryController.cpp (hunk 9 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: reverted

File: src/controller/CommissionerDiscoveryController.cpp (hunk 11 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: addressed

File: src/controller/CommissionerDiscoveryController.cpp (hunk 12 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: added checks and assert to ensure we have the lock

File: src/controller/CommissionerDiscoveryController.cpp (hunk 13 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: added this assert

File: src/controller/CommissionerDiscoveryController.cpp (hunk 15 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: added return check and removed this temporary passcode object

File: src/controller/CommissionerDiscoveryController.cpp (hunk 16 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: added return check and removed this temporary passcode object

File: src/controller/CommissionerDiscoveryController.cpp [hunk 20]
```diff
@@ -94,24 +120,75 @@ void CommissionerDiscoveryController::OnUserDirectedCommissioningRequest(UDCClie
...
+    ChipLogDetail(AppServer, "UX Ok: now on main thread");
+    GetCommissionerDiscoveryController()->InternalOk();
+}
+
+/// Callback for getting execution out of the main chip thread
+void CallbackExternalWorkerOk(System::Layer * aSystemLayer, void * aAppState)
+{
+    ChipLogDetail(AppServer, "UX Ok: out of main thread");
+
+    UDCClientState * client = static_cast<UDCClientState *>(aAppState);
```
[the remark this answers is not in the document]
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: reverted

File: src/controller/CommissionerDiscoveryController.h (hunk 17 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.h:312]: addressed

File: src/controller/CommissionerDiscoveryController.cpp (hunk 18 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: reverted this code

File: src/controller/CommissionerDiscoveryController.cpp [hunk 21]
```diff
@@ -259,8 +473,32 @@ void CommissionerDiscoveryController::Ok()
...
 void CommissionerDiscoveryController::CommissionWithPasscode(uint32_t passcode)
 {
+    uint32_t * tmpPasscode = new uint32_t(passcode);
+    chip::DeviceLayer::StackLock lock;
+    if (CHIP_NO_ERROR !=
+        DeviceLayer::SystemLayer().StartTimer(System::Clock::Seconds32(0), CallbackCommissionWithPasscode,
+                                              static_cast<void *>(tmpPasscode)))
+    {
+        ChipLogError(AppServer, "CommissionWithPasscode: StartTimer returned error");
+        delete tmpPasscode;
```
[the remark this answers is not in the document]
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: reverted this code

File: src/controller/CommissionerDiscoveryController.cpp (hunk 19 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: reverted this code

File: src/controller/CommissionerDiscoveryController.cpp (hunk 2 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: fixed

File: src/controller/CommissionerDiscoveryController.cpp (hunk 10 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: fixed

File: src/controller/CommissionerDiscoveryController.cpp (hunk 14 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp:282]: the logic did not change but in some cases functions had to refactored so that multiple lookups could be done (each with its own callback)

File: src/controller/CommissionerDiscoveryController.cpp (hunk 2 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: I missed that the size comment was referring to the CharSpan. It's true that we don't need the entire max length, we only need the portion through the null termination. Fixed.

**Review by yunhanw-google — APPROVED** [review]

File: src/controller/CommissionerDiscoveryController.cpp [hunk 22]
```diff
@@ -121,86 +140,231 @@ void CallbackOk(System::Layer * aSystemLayer, void * aAppState)
...
+    }
 
-        CharSpan rotatingIdSpan = chip::CharSpan(rotatingIdString, sizeof(rotatingIdString));
-        uint32_t passcode       = 0;
-        uint8_t targetAppCount  = client->GetNumTargetAppInfos();
-        if (targetAppCount > 0)
+    char rotatingIdString[Dnssd::kMaxRotatingIdLen * 2 + 1] = "";
+    Encoding::BytesToUppercaseHexString(client->GetRotatingId(), client->GetRotatingIdLength(), rotatingIdString,
+                                        sizeof(rotatingIdString));
+    CharSpan rotatingIdSpan = CharSpan(rotatingIdString, strlen(rotatingIdString));
```
Comment (bzbarsky-apple) [inline src/controller/CommissionerDiscoveryController.cpp:188]: If we actually want `strlen` here, we should be using `CharSpan::fromCharString`.

But what really confuses me is this: why are we doing any null-termination here at all?  Especially because there is absolutely 0 guarantee that this buffer is in fact null-terminated, because we are never checking for error from BytesToUppercaseHexString.

Why does this code not look like this:

```
char rotatingIdBuffer[Dnssd::kMaxRotatingIdLen * 2];
size_t rotatingIdLength = client->GetRotatingIdLength();
CHIP_ERROR err = Encoding::BytesToUppercaseHexBuffer(client->GetRotatingId(), rotatingIdLength, rotatingIdBuffer,
                                        sizeof(rotatingIdBuffer));
// Handle err being a failure here
CharSpan rotatingIdSpan(rotatingIdBuffer, 2*rotatingIdLength);
```

or so.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/controller/CommissionerDiscoveryController.cpp (hunk 22 above)
Comment (chrisdecenzo) [reply src/controller/CommissionerDiscoveryController.cpp]: fixed
